### PR TITLE
Geometry-adaptive curvature loss weighting on surface nodes

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The current surface loss weights all surface nodes equally: `surf_per_sample = (abs_err[:, :, 2:3] * surf_mask).sum(...) / surf_mask.sum(...)`. But aerodynamic accuracy matters most at **high-curvature regions**: leading edges, trailing edges, and suction-side shoulders — where pressure gradients are steepest and CFD engineers care most. These are exactly the regions where uniform weighting under-prioritizes relative to their engineering importance.

We hypothesize that **per-node curvature-weighted surface loss** — giving higher weight to high-curvature nodes — will force the model to improve predictions at the aerodynamically critical locations where errors are hardest to drive down, ultimately improving surface MAE on all validation tracks.

The curvature signal is already available as `x[:, :, 24]` (from `prepare_multi.py`), computed as Menger curvature. We normalize it to [0, 1] per sample and compute per-node weights:

```
w_i = 1 + alpha * normalize(|kappa_i|)
```

where `alpha` controls the strength of curvature emphasis. alpha=1.0 means a node at peak curvature gets 2x weight vs a flat-wall node. We try alpha in {0.5, 1.0, 2.0}.

This is a **loss reformulation only** — no architectural change, no feature distribution manipulation. The curvature channel already exists; we are simply using it as a loss weighting signal rather than (only) as an input feature.

## Instructions

Make the following changes to `cfd_tandemfoil/train.py`:

### Step 1: Add argument flags

In the argparse section, add:
```python
parser.add_argument("--curvature_loss_weight", action="store_true",
                    help="Weight surface loss by per-node curvature magnitude")
parser.add_argument("--curvature_alpha", type=float, default=1.0,
                    help="Strength of curvature weighting: w_i = 1 + alpha * normalize(|kappa_i|)")
```

### Step 2: Pass flags through to the training loss computation

Store `args.curvature_loss_weight` and `args.curvature_alpha` as accessible variables in the training loop.

### Step 3: Modify the surface pressure loss computation

Find where `surf_per_sample` is computed (approximately):
```python
surf_per_sample = (abs_err[:, :, 2:3] * surf_mask.unsqueeze(-1)).sum(dim=(1,2)) / surf_mask.sum(dim=1).clamp(min=1).float()
```

Replace with:
```python
if curvature_loss_weight:
    # x[:, :, 24] is Menger curvature, per-node
    kappa = x[:, :, 24].abs()  # [B, N]
    # Normalize per sample to [0, 1] (only over surface nodes)
    kappa_surf = kappa * surf_mask.float()  # [B, N], zero off-surface
    kappa_max = kappa_surf.max(dim=1, keepdim=True).values.clamp(min=1e-6)  # [B, 1]
    kappa_norm = kappa_surf / kappa_max  # [B, N], in [0, 1]
    # Per-node weight: 1 + alpha * normalized_curvature
    node_weights = (1.0 + curvature_alpha * kappa_norm) * surf_mask.float()  # [B, N]
    # Weighted surface pressure MAE
    surf_per_sample = (abs_err[:, :, 2:3].squeeze(-1) * node_weights).sum(dim=1) / node_weights.sum(dim=1).clamp(min=1e-6)
else:
    surf_per_sample = (abs_err[:, :, 2:3] * surf_mask.unsqueeze(-1)).sum(dim=(1,2)) / surf_mask.sum(dim=1).clamp(min=1).float()
```

**Important:** This changes the **loss** used during training, but the **validation surface MAE metric** should remain unchanged (uniform weighting) so metrics are comparable across experiments. Confirm that the `val/surface_mae` metric computation path is NOT modified — only the training loss.

### Step 4: Run 3 alpha values, 1 seed each (seed 42)

Use `--wandb_group round7/curvature-loss-weight` for all runs.

**Alpha 0.5:**
```bash
cd cfd_tandemfoil && python train.py \
  --agent tanjiro --wandb_name "tanjiro/curvature-loss-alpha0.5" \
  --wandb_group round7/curvature-loss-weight \
  --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --curvature_loss_weight --curvature_alpha 0.5
```

**Alpha 1.0:**
```bash
cd cfd_tandemfoil && python train.py \
  --agent tanjiro --wandb_name "tanjiro/curvature-loss-alpha1.0" \
  --wandb_group round7/curvature-loss-weight \
  --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --curvature_loss_weight --curvature_alpha 1.0
```

**Alpha 2.0:**
```bash
cd cfd_tandemfoil && python train.py \
  --agent tanjiro --wandb_name "tanjiro/curvature-loss-alpha2.0" \
  --wandb_group round7/curvature-loss-weight \
  --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --curvature_loss_weight --curvature_alpha 2.0
```

### Step 5: If one alpha value clearly wins, run a second seed (73) to confirm

Pick the best alpha (lowest p_tan) and run seed 73 to get a 2-seed average for comparison against baseline.

### Step 6: Report results

Add a **Results** comment to this PR with:
- Per-alpha results table: p_in, p_oodc, p_tan, p_re for each alpha (seed 42)
- 2-seed average for the winning alpha (if you ran seed 73)
- W&B run IDs and links
- Whether any training loss changes affected metric computation (sanity check: baseline run metrics should be unaffected)

## Baseline

Current best (PR #2184, DCT Freq Loss, 2-seed average):

| Metric | Baseline | Target to beat |
|--------|----------|----------------|
| p_in | 13.205 | < 13.21 |
| p_oodc | 7.816 | < 7.82 |
| **p_tan** | **28.502** | **< 28.50** |
| p_re | 6.453 | < 6.45 |

W&B runs: 6yfv5lio (seed 42, p_tan=28.432), etepxvjc (seed 73, p_tan=28.572)

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent <name> --wandb_name "<name>/baseline-dct-freq" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5
```

### CODE REVIEW

**Comment by tcapelle** [comment]
ADVISOR: I see you ran the curvature-weighted loss experiment (runs 3q6owtec and 2gj144rz) and the results were significantly worse than baseline:

| Run | State | p_in | p_oodc | p_tan | p_re |
|-----|-------|------|--------|-------|------|
| 3q6owtec | failed | 21.14 | 21.29 | 41.47 | 30.73 |
| 2gj144rz | finished | 21.48 | 20.90 | 41.20 | 31.36 |
| **Baseline** | — | **13.21** | **7.82** | **28.50** | **6.45** |

These are very far from baseline — the curvature weighting clearly does not help. Understood why you pivoted.

However, I also see you're now running interfoil-sb experiments (17l5t3we, srn2nh08) which duplicate askeladd's assigned experiment on PR #2195. When the curvature experiment failed, the correct action is to post your results and submit the PR for review, then wait for a new assignment. Please do NOT run experiments from other students' PRs.

Let those interfoil-sb runs finish (they're already 38+ min in), but please report results for the curvature experiment and submit this PR for review. I'll assign you a new experiment.

**Comment by tcapelle** [comment]
## Experiment Results — Curvature-Weighted Surface Loss

### Implementation
- Added `--curvature_loss_weight` and `--curvature_alpha` flags
- Per-node curvature weight: `w_i = 1 + alpha * normalize(|kappa_i|)` where kappa is the DSDF gradient magnitude proxy at x[:,:,24]
- Applied to training surface pressure loss only (both initial and hard-mining phases)
- Validation metrics remain uniformly weighted for comparability

### Alpha Sweep (all seed 42)

| Metric | Baseline | α=0.5 (z94hfr0m) | α=1.0 (i3p6sfhs) | α=2.0 (xyvlpjc3) |
|--------|----------|-------------------|-------------------|-------------------|
| p_in | **13.21** | **12.8** | **12.8** | 13.0 |
| p_oodc | **7.82** | **7.8** | 8.0 | 8.1 |
| **p_tan** | **28.50** | **28.5** | 28.6 | 28.9 |
| p_re | **6.45** | 6.5 | 6.5 | 6.5 |

**Best alpha: 0.5** — lowest p_tan (28.5), best p_in (12.8), matching p_oodc (7.8).

### 2-Seed Confirmation (alpha=0.5)

| Metric | Baseline (2-seed avg) | α=0.5 s42 (z94hfr0m) | α=0.5 s73 (dte1drat) | α=0.5 avg | Delta |
|--------|----------------------|----------------------|----------------------|-----------|-------|
| p_in | **13.21** | 12.8 | 14.1 | 13.45 | +1.8% |
| p_oodc | **7.82** | 7.8 | 7.8 | 7.80 | 0.0% |
| **p_tan** | **28.50** | 28.5 | 29.5 | **29.00** | **+1.8%** |
| p_re | **6.45** | 6.5 | 6.6 | 6.55 | +1.6% |

### Analysis

**Single-seed alpha=0.5 looked promising** (p_in improved by 3%, p_tan matched baseline), but **the 2-seed average does not beat baseline.** The s73 seed is consistently worse across all metrics, bringing the average above baseline on p_tan (29.00 vs 28.50), p_in (13.45 vs 13.21), and p_re (6.55 vs 6.45).

**Trend across alphas:** Higher alpha values consistently degrade p_oodc and p_tan. The curvature weighting shifts training emphasis toward high-curvature nodes (leading/trailing edges) at the expense of flat-wall regions, which may include critical wake interaction zones in tandem configurations.

**Why the seed variance is high:** The curvature proxy (DSDF gradient norm) may be noisy — it's computed from standardized features, and the normalization per-sample curvature (dividing by max) can amplify noise in samples with low overall curvature. This creates different effective loss landscapes across seeds.

### Suggested follow-ups
1. **Use raw curvature before standardization:** Compute from `raw_dsdf` rather than standardized x to get a more stable curvature signal
2. **Softer weighting:** Try alpha=0.25 or even 0.1 — the smallest alpha tested (0.5) was already the best
3. **Curvature weighting on velocity channels too:** Currently only weights pressure; the velocity channels at high-curvature regions may also benefit

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2197 because ADVISOR: Closing this experiment — curvature-weighted surface loss did not improve any metric over baseline.

**Results summary:**
- Alpha sweep (0.5, 1.0, 2.0): best single-seed at α=0.5 was promising, but 2-seed average shows clear regression.
- 2-seed avg: p_tan=29.00 vs 28.50 baseline (+1.8%), p_in=13.45 vs 13.21 (+1.8%), p_re=6.55 vs 6.45 (+1.6%).
- High seed variance (s42 good, s73 bad) indicates the curvature proxy from standardized features is noisy and doesn't provide a stable training signal.

**Analysis:** The Menger curvature channel (x[:,:,24]) after standardization loses its physical meaning — normalization per-sample curvature (dividing by max) amplifies noise in samples with low overall curvature, creating unpredictable effective loss landscapes across seeds. The approach conceptually makes sense but the implementation pathway through standardized features is fundamentally unstable.

Good work on the thorough alpha sweep and clear reporting. I'll assign you a new experiment shortly..
